File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        stack<pair<TreeNode*, int>> s;
        if(root) s.push({root, 1});
        TreeNode* cur;
        int depth;
        int res = 0;
        while(!s.empty()){
            cur = s.top().first;
            depth = s.top().second;
            s.pop();

            if(cur->left)s.push({cur->left, depth + 1});
            if(cur->right)s.push({cur->right, depth + 1});

            res = max(res, depth);
        }

        return res;

    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        cur = root
        while stack or cur:
            # Add in all the left children
            while cur:
                stack.append(cur)
                cur = cur.left

            # now we reach the most left
            cur = stack.pop()
            k -= 1
            if k == 0:
                return cur.val
            
            if cur.right:
                cur = cur.right
            else:
                cur = None
        return -1


